### PR TITLE
[Concurrency] Further prevent crashes in legacy mode of isSameExecutor

### PR DESCRIPTION
**Pull request by ktoso**
This cleans up interactions with the "legacy mode" where we're not allowed to crash in the `isCurrentExecutor` impl.

Also cleans up interactions with the SWIFT_UNEXPECTED_EXECUTOR_LOG_LEVEL mode.

This PR also makes all tests to run in both the swift6 and legacy modes, so we have a better overview what's expected.

resolves rdar://128425368
resolves rdar://127400013
resolves rdar://128508780
resolves rdar://128508874

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Repalces https://github.com/apple/swift/pull/73773

**Comment by ktoso** [comment]
This still isn't right I think... doing more investigating and adding more tests for various cases.

File: stdlib/public/Concurrency/Actor.cpp [hunk 1]
```diff
@@ -353,17 +354,15 @@ static void checkIsCurrentExecutorMode(void *context) {
...
-      if (strcmp(modeStr, "nocrash") == 0) {
-        useLegacyMode = true;
-      } else if (strcmp(modeStr, "crash") == 0) {
-        useLegacyMode = false;
-      } // else, just use the platform detected mode
-    }
+  if (const char *modeStr = runtime::environment::concurrencyIsCurrentExecutorLegacyModeOverride()) {
+    if (strcmp(modeStr, "nocrash") == 0 || strcmp(modeStr, "legacy") == 0) {
+      useLegacyMode = true;
+    } else if (strcmp(modeStr, "crash") == 0 || strcmp(modeStr, "swift6") == 0) {
```
Comment (ktoso) [inline stdlib/public/Concurrency/Actor.cpp:360]: just to make it less confusing in our tests; this is not expected to be used by anyone tbh, but I keep the old mode names.

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Review by mikeash — APPROVED** [review]

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Final commit to fix a test on linux/windows, impl remained the same.

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Ofc whack-a-mole with tests in CI now heh

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Pull request merged**